src/compas_ags/ags/graphstatics.py

```diff
@@ -337,6 +351,82 @@ def form_update_from_force(form, force, kmax=100):
         attr['l'] = forces[index, 0]
 
 
+def form_update_from_force_newton(form, force, constraints=None, tol=1e-10, max_iter=20):
+    r"""Update the form diagram after a modification of the force diagram.
+
+    Compute the geometry of the form diagram from the geometry of the force diagram
+    and some constraints, including non-linear.
+    The form diagram is computed by formulating the root-finding approach presented
+    in bi-dir AGS [1]_. Newton's method is used to solve for the form diagram.
+
+    The algorithm fails if it is over-constrained or if the initial guess is too far
+    from any root.
+
+    Parameters
+    ----------
+    form: :class:`FormDiagram`
+        The form diagram to update.
+    force : :class:`ForceDiagram`
+        The force diagram containing the vertex modification desired.
+    constraints : :class: ConstraintsCollection (None)
+        A collection of form diagram constraints.
+    tol: float (1e-10)
+        Stopping criteria tolerance.
+    max_iter : int (20)
```

Comment:
to space or not to space

Reply:
default values should not be provided like this because the rendered version will look awkward.
```
tol: :obj:`float`, optional
    Stopping criterion. Default is ``1e-10``.

Reply:
ok, will update these